File: Algorithm/200608/342.Power of Four.c

// solution 1
bool isPowerOfFour(int num){
    if (num <= 0) return false;
    while (num) {
        if (num & 3) return (num == 1);
        num >>= 2;
    }
    return false;
}

// solution 2
bool isPowerOfFour(int num){
    return ((num > 0) && ((!((num) & (num-1))) && (!((num) & 0xAAAAAAAA))));
}

File: Algorithm/342. Power of Four.py
"""
https://leetcode.com/problems/power-of-four/

Given an integer n, return true if it is a power of four. Otherwise, return false.

An integer n is a power of four, if there exists an integer x such that n == 4x.

 

Example 1:

Input: n = 16
Output: true
Example 2:

Input: n = 5
Output: false
Example 3:

Input: n = 1
Output: true
"""

class Solution:
    def isPowerOfFour(self, n: int) -> bool: 
        if n == 1:
            return True
        if n <= 0:
            return False
        
        while n > 4:
            if n % 4 > 0:
                return False
            n = int(n / 4)
            
        return n == 4
      
      
